### ISSUE TITLE
Repair file size at the end of download if it did not match database entry

### ISSUE THREAD

**Issue by PVince81** [owncloud/core#21751]
In some cases, especially with older encryption migration issues, the file size in the database doesn't match the real file size.

However, for encrypted files, the size cannot be inferred from the encrypted file size until the file has been decrypted at least once.

It is not possible to repair such files with OCC because the password is not known.

So, here is a proposed solution, enhance the download code to do the following:
- after streaming the whole file to the client, take note of how many bytes were sent
- if the number of bytes sent did not match the earlied content-length that was sent => it means that the size did not match
- if the size did not match, update it in the database

There are several code paths to look at:
- web UI download (pre-OC 9.0 where we use Webdav)
- web UI zip download?
- Webdav download

Goal is to have backportable code to be able to fix older installs "on the fly".

@icewind1991 @schiesbn @MorrisJobke thoughts ?

**Comment by schiesbn** [comment]
as said in IRC, I think that's the right way. Maybe we need a third intermediate step if the client stops the download to early if the received size exceeded the content-length the server provided.

I could think about something like:

1. start download with given content-legnth (size stored in db)
2. download fails (if the server read the whole file we can update the size directly but change are high the the client stopped the download to early)
3. set size=-1 in the file cache. This indicates that we don't know the size.
4. The next download will be initiated without any content-length
5. download will work because the client will just download the file to the end
6. Now the server knows the unencrypted size and can update the size column.

This would not only fix the encryption issue but would make the download in general more robust against a broken file cache and would provide a self healing process completely independent from the storage back-end, encryption or any encryption module.

**Comment by PVince81** [comment]
Good point.

So there are two cases to consider/test:
- when size > actual_size: client will wait for more data
- when size < actual_size: client might close the connection early and kill the PHP process (php-fpm)

But then if the connection is killed there isn't much chance to set the size to -1. Hmmm...

In the case of encryption problems, it is more likely that size > actual_size because for some reason size is containing the encrypted file size, which is always bigger.

Let's look into both, but "size > actual_size" has a higher priority.

**Comment by PVince81** [comment]
- [ ] could also happen with version files (another code path)

**Comment by PVince81** [comment]
Looking at the current code, the old ajax/download.php seems to call `$view->filesize($path)` and uses the size returned by that call, which goes through `basicOperation('filesize')`. Looks like in the case without encryption it would just use whichever size is on disk.

So far I tested without encryption, and messed up the size in the database. The returned file through ajax/download.php still has the correct size because the size seems to come directly from the storage.
However through Webdav it has the wrong size.

That's on master / 9.0... will test 8.2.2 later and also with encryption.

**Comment by PVince81** [comment]
POC here https://github.com/owncloud/core/pull/22003

**Comment by PVince81** [comment]
- "occ files:scan --all" doesn't change the size of existing files, even if their mtime has changed

**Comment by MorrisJobke** [comment]
What I did:

* enabled encryption in a 8.2.2
* uploaded a file `welcome-2.txt` (164 Bytes in size)
* change the file size in oc_filecache for this file to `10164`
* :white_check_mark: `curl http://localhost/master/remote.php/webdav/welcome-4.txt -u mjob:mjob -i`:
```
....
Content-Length: 163
...
```
* :x: `curl http://localhost/master/index.php/apps/files/ajax/download.php\?dir\=%2F&files\=welcome-4.txt  -i -u mjob:mjob`
```
...
Content-Length: 10163
...

Welcome to your ownCloud account!

This is just an example file for developers and git users. 
The packaged and released versions will come with better examples.

curl: (18) transfer closed with 10000 bytes remaining to read
```


So WebDAV somehow seems to fix the issue on the fly, while our download.php doesn't.

**Comment by MorrisJobke** [comment]
Doing all this other a share doesn't change anything here.

**Comment by MorrisJobke** [comment]
I increased this to a sev1 because it causes indirect data loss (users can't access files even if they are fine)

**Comment by PVince81** [comment]
The current fix isn't suitable for Webdav because SabreDAV does a simple `file_put_contents($phpOutput, $fileResource)` so we have no control over the loop.

I'll see if it can be moved to the storage wrapper layer. Maybe the encryption storage wrapper can compute the size on the fly and detect if it mismatches at the end.

**Comment by PVince81** [comment]
Here's a much better solution that covers all download/stream_read cases (web UI, Webdav), but very specific to encryption issues: https://github.com/owncloud/core/pull/22008

**Comment by PVince81** [comment]
Related issue, the file size can be wrong when restoring single files from backup: https://github.com/owncloud/core/issues/22096

**Comment by PVince81** [comment]
This would only work if the "encrypted" flag is set to 1.

If someone restores a file from backup and runs "occ files:scan" it will not only have the wrong size but also will have "encrypted" set to 0: https://github.com/owncloud/core/issues/22096

**Comment by PVince81** [comment]
I'd like @schiesbn to have a look at the proposed PR and alternative approaches: https://github.com/owncloud/core/pull/22008

**Comment by PVince81** [comment]
@schiesbn any update on this ?

**Comment by PVince81** [comment]
A much better solution from @schiesbn and @jknockaert here: https://github.com/owncloud/core/pull/22579

**Comment by schiesbn** [comment]
Will be fixed in 9.0.0, 8.1.6 and 8.2.3

**Comment by MorrisJobke** [comment]
@schiesbn @PVince81 @jknockaert Thanks a lot for all your effort to this! :) You all rock! :metal:
